### PR TITLE
Poses mapped

### PR DESCRIPTION
**Pull request by ndan029**
plot poses by pose source (different hard-coded color per pose source)
update our custom pose angular model to include an id
add "id_gt" filter to pose api endpoinnt
create pose api angular service (recursively retrieve poses with optional trial and optional id_gt parameter)

### CODE REVIEW

File: angular/app/src/app/shared/components/maps/map-instance/map-instance.component.ts [hunk 1]
```diff
@@ -555,6 +580,69 @@ export class MapInstanceComponent implements OnChanges, AfterViewInit, OnDestroy
...
+        "features": listOfFeatures,
+      },
+    }
+
+    Object.keys(newPoses).forEach((pose_source_type) => {
+      this.poseSourceLayersSet.add(pose_source_type);
+      Object.keys(newPoses[pose_source_type]).forEach(entity => {
+        // let result = newPoses[pose_source_type][entity].map(a => a.coordinates);
+        newPoses[pose_source_type][entity].forEach(single_pose => {
+          let asdf = {
```
Comment (hohumsup) [inline angular/app/src/app/shared/components/maps/map-instance/map-instance.component.ts]: rename asdf to something more explicit so we know what this variable is used for

File: angular/app/src/app/shared/components/maps/map-instance/map-instance.component.ts [hunk 2]
```diff
@@ -14,7 +14,28 @@ import { FontAwesomeUnicode} from './font-awesome-map'
 import { TimelineCardComponent } from '../../timeline/timeline-card/timeline-card.component';
 import { EventApiService } from 'src/app/shared/services';
 
-
+let COLORS = [
```
Comment (hohumsup) [inline angular/app/src/app/shared/components/maps/map-instance/map-instance.component.ts:17]: thoughts on adding something to similar to `point_style` on` event_types` to `pose_source`? so we can avoid hardcoded colors and any other potential styling changes in the future like shape and color

File: angular/app/src/app/shared/components/maps/map-instance/map-instance.component.ts [hunk 3]
```diff
@@ -555,6 +580,69 @@ export class MapInstanceComponent implements OnChanges, AfterViewInit, OnDestroy
...
+          'type': 'circle',
+          'source': 'posesByPoseSource',
+          'filter': ['==', ['get', 'pose_source'], pose_source],
+          paint: {
+            'circle-radius': 5,
+            'circle-color': COLORS[index],
+          },
+        });
+      }
+    }, this);
```
Comment (hohumsup) [inline angular/app/src/app/shared/components/maps/map-instance/map-instance.component.ts:644]: some poses disappear whenever there's a new pose. however, refreshing the page will display all the poses again.

see below

before (on a refresh)
before_pose

after (a new pose  was posted to the endpoint)
after_pose

in addition to mission poses, the ship icon should be on the furthest right dot but it doesn't get displayed properly unless the page gets refreshed

- [x] fix disappearing poses

File: angular/app/src/app/shared/components/maps/map-instance/map-instance.component.ts [hunk 4]
```diff
@@ -687,6 +775,9 @@ export class MapInstanceComponent implements OnChanges, AfterViewInit, OnDestroy
   }
 
   ngOnChanges(changes: SimpleChanges) {
+    if(changes.posesByPoseSource && this.posesByPoseSource){
+      this.loadPosesSource(changes.posesByPoseSource.currentValue);
```
Comment (hohumsup) [inline angular/app/src/app/shared/components/maps/map-instance/map-instance.component.ts]: here are more info on disappearing poses. below is the output of `changes.posesByPoseSource.currentValue`. it seems like some entity from the pose source gets dropped?

before (on a refresh)
```
GPS: {usv1: Array(6), target_usv: Array(5), map_marker: Array(3)}
Reported: {target_usv: Array(2), map_marker: Array(1), usv1: Array(1)}
```
after (a new pose was posted to the endpoint)
```
GPS: {usv1: Array(7)}
Reported: {target_usv: Array(2), map_marker: Array(1), usv1: Array(1)}
```

File: angular/app/src/app/shared/components/maps/maps.component.ts [hunk 5]
```diff
@@ -43,25 +43,55 @@ export class MapsComponent implements OnInit, OnDestroy {
...
+    this.subscriptions.add(_poseApiService.getPoses().subscribe(
+      many_poses =>  {
+        // make a shallow copy to reassign posesByPoseSource later
+        let new_mapping = Object.assign({}, this.posesByPoseSource);
+        Object.keys(many_poses).forEach(pose_source => {
+          let pose_source_mapping = {};
+          Object.keys(many_poses[pose_source]).forEach(entity => {
+            let new_list;
+            if(this.posesByPoseSource.hasOwnProperty(pose_source) && this.posesByPoseSource[pose_source][entity]) {
+              new_list =  [...this.posesByPoseSource[pose_source][entity], ...many_poses[pose_source][entity]];
```
Comment (hohumsup) [inline angular/app/src/app/shared/components/maps/maps.component.ts:81]: i think this is where some poses get dropped when there's a new pose

File: angular/app/src/app/shared/services/map/pose-api.service.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  // retrieve poses using django cursor pagination url
+  private getPosesContinued(url: string) {
+    let posesRequest: Observable<PosePageResult> = this.performSearch(url)
+    this.recursiveSubscription = posesRequest.subscribe(this.catagorizePoses.bind(this));
+  }
+
+  // sort poses by pose source, then by entity
+  catagorizePoses(page: PosePageResult) {
```
Comment (hohumsup) [inline angular/app/src/app/shared/services/map/pose-api.service.ts]:
```suggestion
  categorizePoses(page: PosePageResult) {
```

File: angular/app/src/app/shared/services/map/pose-api.service.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    };
+    return this.http.get(url, options).pipe(
+      map((data: any) => PoseAdapters.posePageResultAdapter(data) ),
+    );
+  }
+
+  // retrieve poses using trial id
+  private getPosesBase(): void {
+    let posesRequest: Observable<PosePageResult> = this.performSearch('/api/poses/', this.mostRecentPoseID, this.selectedTrialId)
+    this.recursiveSubscription = posesRequest.subscribe(this.catagorizePoses.bind(this));
```
Comment (hohumsup) [inline angular/app/src/app/shared/services/map/pose-api.service.ts]:
```suggestion
    this.recursiveSubscription = posesRequest.subscribe(this.categorizePoses.bind(this));
```

File: angular/app/src/app/shared/services/map/pose-api.service.ts [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  // retrieve poses using django cursor pagination url
+  private getPosesContinued(url: string) {
+    let posesRequest: Observable<PosePageResult> = this.performSearch(url)
+    this.recursiveSubscription = posesRequest.subscribe(this.catagorizePoses.bind(this));
```
Comment (hohumsup) [inline angular/app/src/app/shared/services/map/pose-api.service.ts]:
```suggestion
    this.recursiveSubscription = posesRequest.subscribe(this.categorizePoses.bind(this));
```

**Review by hohumsup — CHANGES_REQUESTED** [review]

File: angular/app/src/app/shared/components/maps/map-instance/map-instance.component.ts (hunk 4 above)
Comment (ndan029) [reply angular/app/src/app/shared/components/maps/map-instance/map-instance.component.ts]: Does this happen on the angular dev server? I had this happen to me and I thought I had fixed it.

Comment (hohumsup) [reply angular/app/src/app/shared/components/maps/map-instance/map-instance.component.ts]: yeah, i saw the error occurring on the dev server.

File: angular/app/src/app/shared/components/maps/map-instance/map-instance.component.ts (hunk 3 above)
Comment (ndan029) [reply angular/app/src/app/shared/components/maps/map-instance/map-instance.component.ts:644]: Fixed

**Review by hohumsup — APPROVED** [review]
map changing bug in another PR

**Pull request closed without merging**